src/Pate/Verification/Widening.hs

```diff
@@ -171,31 +179,123 @@ widenAlongEdge bundle from d gr to = withSym $ \sym ->
           WideningError msg _ d'' ->
             do let msg' = ("Error during widening: " ++ msg)
                traceBundle bundle msg'
-               postSpec' <- abstractOverVars bundle from to postSpec d''
+               postSpec' <- abstractOverVars scope bundle from to postSpec d''
                case updateDomain gr from to postSpec' of
                  Left gr' ->
                    do traceBundle bundle ("Ran out of gas while widening postconditon! " ++ show from ++ " " ++ show to)
                       return $ recordMiscAnalysisError gr' to (Text.pack msg')
                  Right gr' -> return $ recordMiscAnalysisError gr' to (Text.pack msg')
 
           Widen _ _ d'' -> do
-            postSpec' <- abstractOverVars bundle from to postSpec d''
+            postSpec' <- abstractOverVars scope bundle from to postSpec d''
             case updateDomain gr from to postSpec' of
               Left gr' ->
                 do traceBundle bundle ("Ran out of gas while widening postconditon! " ++ show from ++ " " ++ show to)
                    return gr'
               Right gr' -> return gr'
 
--- TODO: implement this to actually properly perform the scope transformation
+-- | Compute an'PAD.AbstractDomainSpec' from the input 'PAD.AbstractDomain' that is
+-- parameterized over the *output* state of the given 'SimBundle'.
+-- Until now, the widening process has used the same scope for the pre-domain and
+-- post-domain (i.e. both contain free variables corresponding to the initial values
+-- before symbolic execution).
+-- To abstract the computed post-domain from its calling context, we need to rephrase
+-- any symbolic terms it contains so that they only refer to the output state.
+-- Specifically, given a post-domain @dom[pre]@ phrased over the pre-state variables, and
+-- a function @f(pre)@ representing the result of symbolic execution, we want to compute
+-- @dom'[post]@ such that @dom'[post/f(pre)] == dom[pre]@.
```

Comment:
Can you add an explanation here that there are multiple strategies for accomplishing this (i.e., that list of `Alternative`s at the end)? I don't think you need to explain each one in detail, but that there are multiple and at least one is expected to work for each case.

Reply:
Added comment